### PR TITLE
update nemo rl command

### PR DESCRIPTION
**Pull request by wedu-nvidia**

### CODE REVIEW

**Comment by wedu-nvidia** [comment]
Just put draft version here and I will double check before mark it as ready

**Comment by wedu-nvidia** [comment]
I haven’t completed full loop testing for some parts of the code yet, but I wanted to bring more people in to help review in case I missed any important changes.

File: docs/pipelines/training-verl.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        '++actor_rollout_ref.rollout.tensor_model_parallel_size=2 '
+    ),
+    cluster="slurm",
+    expname="test-verl-ppo",
+    output_dir="/workspace/test-verl-ppo",
+    hf_model="/hf_models/Qwen2.5-1.5B-Instruct",
+    prompt_data="/data/rl-data.parquet",
+    num_gpus=8,
+    num_nodes=2,
+    # this is used for the LLM judge
```
Comment (Kipok) [inline docs/pipelines/training-verl.md]: just wondering is this command still functional? If not, might be good to update to something that works (e.g. not using judge)

File: docs/pipelines/training.md [hunk 2]
```diff
@@ -41,200 +39,103 @@ a good idea to check their documentation to understand how this config is struct
...
+
+We also support sequence packing and context parallel, espeically for training sequences > 4k or so, it's recommended to use sequence packing and context parallel.
+(By default, our sft config set sequence_packing as True)
+```bash
+   ++policy.sequence_packing.enabled=True \
+   ++policy.context_parallel_size=4
+```
+
+
+The training script will automatically convert the final saved checkpoint into the Hugging Face format and average all generated checkpoints upon completion (**checkpoints average TBD**)).
```
Comment (Kipok) [inline docs/pipelines/training.md]: if it's not really happening, let's not write it in the docs. We can update it later when we support averaging

File: docs/pipelines/training.md [hunk 3]
```diff
@@ -41,200 +39,103 @@ a good idea to check their documentation to understand how this config is struct
...
 output_dir = f"/workspace/{expname}/checkpoints"
 
-train(
+sft_nemo_rl(
     ctx=wrap_arguments(""),
     cluster=cluster,
     expname=expname,
     output_dir=output_dir,
-    nemo_model="/nemo_models/llama3.1-8b-base",
+    hf_model="/hf_models/llama3.1-8b-base",
```
Comment (Kipok) [inline docs/pipelines/training.md]: Let's reference model as meta-llama/... so that it's automatically downloaded

Comment (Kipok) [reply docs/pipelines/training.md]: Let's do it in all relevant commands

File: docs/releases/openmathinstruct2/training.md [hunk 4]
```diff
@@ -49,73 +49,66 @@ python -m nemo_skills.training.prepare_data \
...
 
+
 If you have a job timeout, it's necessary to set the maximum time per run to 40 minutes
 before the timeout to allow for the final checkpoint to be saved. E.g. if your timeout is 4 hours,
-add `++exp_manager.max_time_per_run=00:03:20:00`
+add `++checkpointing.checkpoint_must_save_by=00:03:20:00`
 
 
-If you want to follow up with checkpoint conversion and evaluation, see
+If you want to follow up with evaluation, see
```
Comment (Kipok) [inline docs/releases/openmathinstruct2/training.md]: Please add a note here that in the original work we did checkpoint averaging, but it's not done in the current code and results might be a bit different. Can also say that we will add the support soon

File: docs/releases/openmathreasoning/training.md [hunk 5]
```diff
@@ -155,49 +82,38 @@ sed -i 's/"max_position_embeddings": 4096,/"max_position_embeddings": 131072,/g'
...
+    ++checkpointing.save_period=7500 \
+    ++sft.max_num_steps=30000 \
+    ++sft.max_num_epochs=100
 ```
 
-Note that while we set batch size to be 256, the *real* batch size is about 4 times bigger as there are approximately
-4 examples packed together in each element of the packed data.
+
 
 For other models change the above parameters according to this table. Don't forget to re-pack the data when changing CP!
```
Comment (Kipok) [inline docs/releases/openmathreasoning/training.md]: Remove last sentence as re-packing isn't relevant

File: docs/tutorials/posts/omr-simple-recipe.md [hunk 6]
```diff
@@ -191,43 +191,8 @@ ns run_cmd --log_dir=/workspace/prepare-sft-data --expname=prepare-sft-data --ru
...
-    ++model.data.train_ds.max_seq_length=8192 \
-    ++model.data.train_ds.global_batch_size=32 \
-    ++model.tensor_model_parallel_size=4 \
-    ++model.context_parallel_size=2 \
-    ++model.optim.lr=1e-5 \
-    ++trainer.sft.max_epochs=2
-```
-
-For the NeMo-RL backend, use the following training command. Add `--disable_wandb` to disable W&B logging. Only run one of the training commands, not both (or change the paths and `expnames` accordingly).
+For the NeMo-RL with megatron backend, use the following training command. Add `--disable_wandb` to disable W&B logging. Only run one of the training commands, not both (or change the paths and `expnames` accordingly).
```
Comment (Kipok) [inline docs/tutorials/posts/omr-simple-recipe.md]: remove last sentence

File: docs/tutorials/posts/omr-simple-recipe.md [hunk 7]
```diff
@@ -258,23 +223,10 @@ To learn more about SFT configuration, see the [NeMo-Skills training](https://nv
 To check model improvement, let's run another evaluation. Convert the checkpoint back into Hugging Face format for faster evaluation. You can skip the conversion step if you’re using the NeMo-RL backend for training.
```
Comment (Kipok) [inline docs/tutorials/posts/omr-simple-recipe.md]: remove last sentence

File: docs/index.md [hunk 8]
```diff
@@ -23,7 +23,7 @@ Here are some of the features we support:
         - [**Tool-calling**](./evaluation/tool-calling.md): e.g. [bfcl_v3](./evaluation/tool-calling.md#bfcl_v3)
         - [**Robustness Evaluation**](./evaluation/robustness.md): Evaluate model sensitvity against changes in prompt.
     - Easily parallelize each evaluation across many Slurm jobs, self-host LLM judges, bring your own prompts or change benchmark configuration in any other way.
-- [Model training](pipelines/training.md): Train models using [NeMo-Aligner](https://github.com/NVIDIA/NeMo-Aligner/), [NeMo-RL](https://github.com/NVIDIA/NeMo-RL/) or [verl](https://github.com/volcengine/verl).
+- [Model training](pipelines/training.md): Train models using [NeMo-RL](https://github.com/NVIDIA-NeMo/RL/), or [verl](https://github.com/volcengine/verl).
```
Comment (Kipok) [inline docs/index.md]:
```suggestion
- [Model training](pipelines/training.md): Train models using [NeMo-RL](https://github.com/NVIDIA-NeMo/RL/) or [verl](https://github.com/volcengine/verl).
```

File: docs/pipelines/training-verl.md (hunk 1 above)
Comment (wedu-nvidia) [reply docs/pipelines/training-verl.md]: Ok, I can have a check for this, and update if needed

File: docs/pipelines/training.md (hunk 2 above)
Comment (wedu-nvidia) [reply docs/pipelines/training.md]: sure

**Comment by Kipok** [comment]
@wedu-nvidia please fix conflicts and let's aim to merge this tomorrow

File: docs/pipelines/training.md [hunk 9]
```diff
@@ -32,209 +32,107 @@ that data preparation script supports which you can see
...
 
-We use [NeMo-Aligner](https://github.com/NVIDIA/NeMo-Aligner/) to run LLM training,
+We use [NeMo-RL](https://github.com/NVIDIA-NeMo/RL) to run LLM training,
 so you can check their documentation to learn about all supported parameters.
 
 Here is an example of how to run a training job.
+It supports models in the Hugging Face format directly.
+
+In the example below, we use Megatron as the backend.
+NeMo-RL provides two configuration options: fsdp and megatron
```
Comment (Kipok) [inline docs/pipelines/training.md]:
```suggestion
NeMo-RL supports two training backends: fsdp and megatron. fsdp is typically slower but supports more models. If your model is supported by megatron we recommend using it.
```

File: docs/pipelines/training.md [hunk 10]
```diff
@@ -32,209 +32,107 @@ that data preparation script supports which you can see
...
     --training_data=/data/sft-data.jsonl
 ```
 
 This will run training on 8 nodes of 8 GPUs, using 4 dependent slurm jobs.
-By default we are training for 2 epochs, saving checkpoints every 1000 steps,
-but you can adjust these values. It's also recommended to tune micro batch size
-and tensor parallel parameters for optimal performance. E.g. these are good
-defaults for an 8B model size
+You can adjust the number of epochs and steps as shown below.
+The training will finish earlier once either the specified number of epochs or steps is reached.
```
Comment (Kipok) [inline docs/pipelines/training.md]:
```suggestion
The training will finish once either the specified number of epochs or steps is reached.
```

File: docs/pipelines/training.md [hunk 11]
```diff
@@ -32,209 +32,107 @@ that data preparation script supports which you can see
...
+
+```bash
+    ++sft.max_num_epochs=2 \
+    ++sft.max_num_steps=1000 \
+```
+
+It is also recommended to tune the micro batch size and tensor parallel parameters for optimal performance.
```
Comment (Kipok) [inline docs/pipelines/training.md]:
```suggestion
It is also recommended to tune the micro batch size, max sequence length and parallelism parameters for optimal performance. If sequence packing is enabled (default) it's recommended to keep micro batch size as 1 and instead increase sequence packing length when GPU memory isn't used fully.
```

File: docs/pipelines/training.md [hunk 12]
```diff
@@ -32,209 +32,107 @@ that data preparation script supports which you can see
...
+
+For dense models (e.g., Qwen3-8B), adjusting these settings can significantly improve training efficiency.
 
 ```bash
-    ++model.data.train_ds.micro_batch_size=4 \
-    ++model.tensor_model_parallel_size=4
+    ++policy.train_global_batch_size=32 \
+    ++policy.train_micro_batch_size=1 \
+    ++policy.tensor_model_parallel_size=4
```
Comment (Kipok) [inline docs/pipelines/training.md]:
```suggestion
    ++policy.tensor_model_parallel_size=4 \
```

File: docs/pipelines/training.md [hunk 13]
```diff
@@ -32,209 +32,107 @@ that data preparation script supports which you can see
...
 ```
 
-The training script will average all of your generated checkpoints upon completion
-(we found this to consistently increase the downstream accuracy). If you want to
-only average a subset of checkpoint, add `--average_steps` parameter (e.g. if you
-want to disable averaging, set it to the last training step). If you only want
-to average the checkpoints of the finished job, set `--num_training_jobs=0`.
+
+We also support sequence packing and context parallel, especially for training sequences > 4k or so, it's recommended to use sequence packing and context parallel.
+(By default, our sft config set sequence_packing as True)
```
Comment (Kipok) [inline docs/pipelines/training.md]:
```suggestion
By default, our sft config set sequence_packing as True.
```

File: docs/pipelines/training.md [hunk 14]
```diff
@@ -32,209 +32,107 @@ that data preparation script supports which you can see
...
+```bash
+   ++policy.sequence_packing.enabled=True \
+   ++policy.context_parallel_size=4
+```
+
+
+The training script will automatically convert the final saved checkpoint into the Hugging Face format.
```
Comment (Kipok) [inline docs/pipelines/training.md]:
```suggestion
The training script will automatically convert the final saved checkpoint into the HuggingFace format.
```

File: docs/pipelines/training.md [hunk 15]
```diff
@@ -32,209 +32,107 @@ that data preparation script supports which you can see
...
+If you want to only average a subset of checkpoint, add `--average_steps` parameter (e.g. --average_steps=100,200,300).
```
Comment (Kipok) [inline docs/pipelines/training.md]:
```suggestion
If you want to convert non-final checkpoint, use `--conversion_step=X`. If you want to average a subset of checkpoints, add `--average_steps` parameter (e.g. --average_steps=100,200,300). 
```

File: docs/releases/openmathinstruct2/training.md [hunk 16]
```diff
@@ -49,73 +49,67 @@ python -m nemo_skills.training.prepare_data \
...
+    ++policy.train_micro_batch_size=8 \
+    ++policy.train_global_batch_size=512 \
+    ++policy.tensor_model_parallel_size=4 \
+    ++policy.pipeline_model_parallel_size=1 \
+    ++policy.lr=2e-5 \
+    ++checkpointing.save_period=10000 \
+    ++sft.max_num_steps=60000 \
+    ++sft.max_num_epochs=100 \
+    ++policy.sequence_packing.enabled=False \
+    ++checkpointing.checkpoint_must_save_by=00:03:45:00
```
Comment (Kipok) [inline docs/releases/openmathinstruct2/training.md]:
```suggestion
```

File: docs/releases/openmathinstruct2/training.md [hunk 17]
```diff
@@ -49,73 +49,67 @@ python -m nemo_skills.training.prepare_data \
...
-    ++trainer.sft.max_epochs=100
+    ++policy.train_micro_batch_size=8 \
+    ++policy.train_global_batch_size=512 \
+    ++policy.tensor_model_parallel_size=4 \
+    ++policy.pipeline_model_parallel_size=1 \
+    ++policy.lr=2e-5 \
+    ++checkpointing.save_period=10000 \
+    ++sft.max_num_steps=60000 \
+    ++sft.max_num_epochs=100 \
+    ++policy.sequence_packing.enabled=False \
```
Comment (Kipok) [inline docs/releases/openmathinstruct2/training.md]:
```suggestion
    ++policy.sequence_packing.enabled=False
```

File: docs/releases/openmathinstruct2/training.md [hunk 18]
```diff
@@ -49,73 +49,67 @@ python -m nemo_skills.training.prepare_data \
...
-    ++trainer.sft.max_epochs=100
+    ++policy.train_micro_batch_size=1 \
+    ++policy.train_global_batch_size=512 \
+    ++policy.tensor_model_parallel_size=8 \
+    ++policy.pipeline_model_parallel_size=2 \
+    ++policy.lr=1e-5 \
+    ++checkpointing.save_period=3330 \
+    ++sft.max_num_steps=20000 \
+    ++sft.max_num_epochs=100 \
+    ++policy.sequence_packing.enabled=False \
```
Comment (Kipok) [inline docs/releases/openmathinstruct2/training.md]:
```suggestion
    ++policy.sequence_packing.enabled=False
```

File: docs/releases/openmathinstruct2/training.md [hunk 19]
```diff
@@ -49,73 +49,67 @@ python -m nemo_skills.training.prepare_data \
...
+    ++checkpointing.checkpoint_must_save_by=00:03:20:00
```
Comment (Kipok) [inline docs/releases/openmathinstruct2/training.md]:
```suggestion

File: nemo_skills/training/verl/math_reward.py
Comment (Kipok) [inline nemo_skills/training/verl/math_reward.py:1]: Let's maybe remove this completely

Comment (wedu-nvidia) [reply nemo_skills/training/verl/math_reward.py:1]: ok, sure.

**Comment by wedu-nvidia** [comment]
@Kipok All done from my side.
